packages/amplify-graphql-relational-transformer/src/data-source-based-directive-transformer.ts

```diff
@@ -0,0 +1,21 @@
+import {
+  TransformerContextProvider,
+  TransformerPrepareStepContextProvider,
+  ModelDataSourceStrategyDbType,
+  TransformerTransformSchemaStepContextProvider,
+} from '@aws-amplify/graphql-transformer-interfaces';
+import { BelongsToDirectiveConfiguration, HasManyDirectiveConfiguration, HasOneDirectiveConfiguration } from './types';
+
+export type DirectiveConfiguration = HasOneDirectiveConfiguration | HasManyDirectiveConfiguration | BelongsToDirectiveConfiguration;
+
+/**
+ * Represents a subset of transformer methods based on a specific
+ * data source (currently SQL / DDB Fields).
+ */
+export interface DataSourceBasedDirectiveTransformer<Config extends DirectiveConfiguration> {
+  dbType: ModelDataSourceStrategyDbType;
+  prepare: (context: TransformerPrepareStepContextProvider, config: Config) => void;
+  transformSchema: (context: TransformerTransformSchemaStepContextProvider, config: Config) => void;
+  generateResolvers: (context: TransformerContextProvider, config: Config) => void;
+  validate: (context: TransformerContextProvider, config: Config) => void;
+}
```

Comment:
One of the goals with this interface and its implementing types is to have the compiler prevent passing an inapplicable directive configuration. If you see a simpler/better way to do this, please let me know!

Reply:
Is it your goal that the interface provides type safety to prevent incorrect directive configs, or that it gives a framework for runtime validation?

Reply:
The former. I want this to be a compiler error:
```ts
const config: HasOneDirectiveConfiguration = ...
// ❌ Argument of type 'HasOneDirectiveConfiguration' is not assignable 
// to parameter of type 'BelongsToDirectiveConfiguration'.
getBelongsToDirectiveTransformer(dbType).prepare(context, config);
```
This PR accomplishes that. But I assume there's a typescripty-er way of doing this -- hence the ask.

Reply:
Thanks for clarifying. I think the approach is reasonable--you know I love me some generics--but we'll want to refine the naming a bit. Even though they aren't publicly exported, we want to be more specific about the intended usages of these types. We could change `DirectiveConfiguration` to `RelationalDirectiveConfiguration` but that leaves open the question of why `ManyToManyDirectiveConfiguration` isn't included. Naming is hard, so I will leave this to you :P

Reply:
I like the `RelationalDirectiveConfiguration` naming suggestion -- addressed in 87fae81efe5d7ff4ed850dc32f598d1a8458e636.

For consistency's sake, we could add a `DataSourceBasedDirectiveTransformer<ManyToManyDirectiveConfiguration>` with a single implementation for all data sources. I'm inclined to leave it as is for now though.